src/dune_rules/melange/melange_rules.ml

```diff
@@ -152,12 +152,40 @@ let js_targets_of_libs sctx libs ~module_systems ~target_dir =
         List.rev_append for_vlib base))
 ;;
 
-let build_js ~loc ~dir ~pkg_name ~mode ~module_systems ~output ~obj_dir ~sctx ~includes m =
+let build_js
+  ~loc
+  ~dir
+  ~pkg_name
+  ~mode
+  ~module_systems
+  ~output
+  ~(obj_dir : Path.t Obj_dir.t)
+  ~sctx
+  ~includes
+  ~local_modules
+  m
+  =
   let open Memo.O in
   let* compiler = Melange_binary.melc sctx ~loc:(Some loc) ~dir in
   Memo.parallel_iter module_systems ~f:(fun (module_system, js_ext) ->
     let src = Obj_dir.Module.cm_file_exn obj_dir m ~kind:(Melange Cmj) in
     let output = make_js_name ~output ~js_ext m in
+    let local_library_paths =
+      match local_modules with
+      | Some (modules, (obj_dir : Path.Build.t Obj_dir.t)) ->
+        Dep_rules.immediate_deps_of m modules ~obj_dir ~ml_kind:Impl
+        |> Action_builder.map ~f:(fun (dep_ms : Module.t list) ->
+          List.map dep_ms ~f:(fun (dep_m : Module.t) ->
+            let path = Obj_dir.Module.cm_file_exn obj_dir dep_m ~kind:(Melange Cmj) in
+            Path.Build.set_extension ~ext:".cmj" path |> Path.build))
+      | None -> Action_builder.return []
+    in
+    let* local_library_paths =
+      let+ local_library_paths, _deps =
+        Action_builder.evaluate_and_collect_deps local_library_paths
```

Comment:
You shouldn't build this here but try to insert it using one of the dyn deps constructor.